akka-stream/src/main/scala/akka/stream/impl/TraversalBuilder.scala

```diff
@@ -133,18 +134,32 @@ object TraversalBuilder {
    */
   private[impl] def initShape(shape: Shape): Unit = {
     // Initialize port IDs
-    val inIter = shape.inlets.iterator
-    var i = 0
-    while (inIter.hasNext) {
-      inIter.next.id = i
-      i += 1
+    val inlets = shape.inlets
+    if (inlets.nonEmpty) {
```

Comment:
I think you can also go after the constructor in `GraphStageLogic` which takes a shape, but it actually just passes the number of input and output ports to the real constrcutror.